[ENHANCEMENT] [REFACTOR] optimise and refactor SDK ingestion methods

burtenshaw opened the pull request:
This PR refactors refactors the ingestion flow in `DatasetRecords` by implementing a new class and module `IngestedRecordMapper` 

This PR supports mapping incoming columns/keys to dataset attributes in these two ways: 

- supports tuple values in the mapping parameter of the log method so that user can specify the two attributes as a tuple.
- refactors in the `_ingest_records` methods so that mapping is performed once before the ingestion loop instead of during.

This PR also optimises the `log` method so that it takes less time and is easier to work with:
- uses tqdm to log status 
- uses exception to show bad records
- iterates over the map not the data

<img width="689" alt="Screenshot 2024-06-26 at 09 09 11" src="https://github.com/argilla-io/argilla/assets/19620375/fbdf2840-1d17-4764-bd4c-2665e95afc14">


- Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**
- tests have been modified, deprecated, and updated to support changes in the ingestion flow

**Checklist**
<!--  Please go over the list and make sure you've taken everything into account -->

- I added relevant documentation
- follows the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

burtenshaw committed:
test: update tests for refactored mapping method

burtenshaw committed:
refactor: introduce independent mapping method and move logic to before ingestion loop

burtenshaw committed:
docs: update all doc strings in dataset records

burtenshaw committed:
chore: improve typing and docs on type

burtenshaw committed:
docs: wrong method in records api reference

burtenshaw committed:
feat: add exception for record ingestion

burtenshaw committed:
refactor: improve explainabilitity and readability in ingestion code with refactor and tqdm and exceptions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

davidberenstein1957 commented on argilla/src/argilla/records/_dataset_records.py:
```diff
@@ -239,7 +241,7 @@ def log(
         records_updated = 0
         for batch in tqdm(
-            iterable=range(0, len(records), batch_size), desc="Adding and updating records", unit="batch"
+            iterable=range(0, len(records), batch_size), desc="2/2: Adding and updating records", unit="batch"
```
```suggestion
            iterable=range(0, len(records), batch_size), desc="Adding and updating records", unit="batch"
```

davidberenstein1957 replied:
Not sure if 1/2 and 2/2 adds a lot or maybe makes it difficult to maintain?

frascuchon replied:
I can imagine that there is a mechanism to handle this progress in `tqdm`. Something like task `i` of `n` progress.

burtenshaw replied:
In fact, since step 1 is now completely local, we might as well skip the progress bar.

davidberenstein1957 commented on argilla/src/argilla/records/_dataset_records.py:
```diff
@@ -398,92 +410,103 @@ def _validate_vector_names(self, vector_names: Union[List[str], str]) -> None:
+        # update the mapping with unmapped columns
+        for key, value in records[0].items():
+            if key not in schema and key not in mapping:
+                warnings.warn(
```
not sure if we would skip in both cases? I think if it is not in the schema, I would say the user made a mistake right? If it it not present in the one record, I would assume it might by chance have an empty value.

burtenshaw replied:
If a value is not present in the schema or mapping, then we assume that it's something that the user does not want to log to Argilla. We warn in case they've made a mistake.

davidberenstein1957 commented on argilla/src/argilla/records/_dataset_records.py:
```diff
@@ -398,92 +410,103 @@ def _validate_vector_names(self, vector_names: Union[List[str], str]) -> None:
+
+                attribute_name = attribute_mapping[0]
+                schema_item = schema.get(attribute_name)
+                attribute_type = attribute_mapping[1] if len(attribute_mapping) > 1 else None
```
don't we want to check if all attribute types are the same and if not can't there be different attribute types? perhaps I want to map to metadata and field or something?

burtenshaw replied:
Sorry, I don't understand this question. The user can map a value in their data by doing:

```python
mapping = {
        "my_prompt": ("prompt_field", "prompt_question"),
    }
```

The strings in the tuple could point to metadata etc.

davidberenstein1957 commented on argilla/src/argilla/records/_dataset_records.py:
```diff
@@ -494,3 +517,52 @@ def _infer_record_from_mapping(
+            value = data.get(source_key)
+            schema_item = self.__dataset.schema.get(attribute_name)
+            suggestion_values[attribute_name].update(
+                {sub_attribute: value, "question_name": attribute_name, "question_id": schema_item.id}
```
```suggestion
                {sub_attribute: value, "question_name": attribute_name, "question_id": schema_item.id}
```
maybe `sub_attribute` is not very specific and can be `attribute_value_key` or something?

burtenshaw replied:
sub_attribute as yet only refers to the parameters of a suggestion. So score and agent. I don't think `_key` explains that, but I'm open to suggestions.

frascuchon replied:
Okay. Now I think about what these methods do. So the idea is to generate a normalized mapping which will be used to build the record. Maybe this mapping can be done once and used for all the records since we have the dataset schema which can be used as a default mapping. I have some comments about the implementation.

1. The `_render_record_mapping` can be executed  just once if we use the settings schema
2. I think it would be easier to follow if the "singular_mapping" is a class instead of a raw dict. This will be a normalized representation of the mapping so we can have something like:

```python
# this will apply the `_render_record_mapping` logic (I guess)
mapping = NormalizedMapping(mapping, schema=settings.schema)

# at some point inside _infer_record_from_mapping
# Parse suggestions and responses into objects aligned with questions
suggestions = self._parse_suggestion_from_mapping(data=data, mapping=mapping.suggestions)
responses = self._parse_response_from_mapping(data=data, mapping=mapping.responses, user_id=user_id)

suggestions = [Suggestion(**suggestion_dict) for suggestion_dict in suggestion_values.values()]
# Parse fields, metadata, and vectors into
fields = {attribute_name: data.get(source_key) for attribute_name, source_key in mapping.fields.items()}
metadata = {attribute_name: data.get(source_key) for attribute_name, source_key in mapping.metadata.items()}
vectors = {attribute_name: data.get(source_key) for attribute_name, source_key in mapping.vectors.items()}
```

or even more simple:
```python
mapping = NormalizedMapping(mapping, schema=settings.schema)
...

suggestions = mapping.suggestions(data)
responses = mapping.responses(data, user_id)
fields = mapping.fields(data)
metadata = mapping.metadata(data)
vectors = mapping.vectors(data)

```
or just in once:
```python
record = mapping.infer_record(data)
```

burtenshaw replied:
@frascuchon I've reimplemented the PR using a class as suggested.

burtenshaw committed:
enhancement: move mapping out of record loop

davidberenstein1957 reviewed (commented):
Left some initial comments.

frascuchon commented on argilla/src/argilla/records/_dataset_records.py:
```diff
@@ -358,25 +360,35 @@ def to_datasets(self) -> HFDataset:
-        return [record.api_model() for record in records]
+
+        ingested_records = []
+        for record in tqdm(records, desc="1/2: Ingesting records", unit="record"):
```
Yes, is a bit weird having this 1/2 in an isolated method without the whole progress picture.
```suggestion
        for record in tqdm(records, desc="Ingesting records", unit="record"):
```

frascuchon commented on argilla/src/argilla/records/_dataset_records.py:
```diff
@@ -358,25 +360,35 @@ def to_datasets(self) -> HFDataset:
+            try:
+                if not isinstance(record, Record):
+                    rendered_mapping = self._render_record_mapping(records=records, mapping=mapping)
+                    record = self._infer_record_from_mapping(data=record, mapping=rendered_mapping, user_id=user_id)
```
I think we should merge both methods in once
```suggestion
                    record = self._infer_record_from_mapping(data=record, mapping=mapping, user_id=user_id)                    
```
and, internally, the `_infer_record_from_mapping` will have a call to                     rendered_mapping = self._render_record_mapping(records=records, mapping=mapping)

burtenshaw replied:
I've updated this method so that we only call `_render_record_mapping` once before the loop. Do you think we should call it every time?

frascuchon commented on argilla/src/argilla/records/_dataset_records.py:
```diff
@@ -208,7 +209,7 @@ def __repr__(self) -> str:
         self,
         records: Union[List[dict], List[Record], HFDataset],
-        mapping: Optional[Dict[str, str]] = None,
+        mapping: Optional[Dict[str, Union[str, List[str], Tuple[str]]]] = None,
```
Why not just support list or tuple instead of both? Or just a Sequence[str]?

frascuchon committed:
[REFACTOR] Avoid autofetch when accessing settings (#5112)

Reviewing and improving records.log

Instead of: 

<img width="1335" alt="Captura de pantalla 2024-06-26 a las 12 48 14"
src="https://github.com/argilla-io/argilla/assets/2518789/02283f4c-fe6a-464f-96b3-36853e6c7622">

for 50 records, records.log can log 1000:

<img width="870" alt="Captura de pantalla 2024-06-26 a las 12 48 57"
src="https://github.com/argilla-io/argilla/assets/2518789/d20f0469-0b33-427e-aa12-b4b7e1d40cd1">

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' of https://github.com/argilla-io/argilla into spike/mapping-to-tuple

burtenshaw committed:
enhancement: use just one progress bar

burtenshaw committed:
chore: update typing of mapping

burtenshaw committed:
fix: move render mapping into infer record method

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
fix: align add records parameters with render function

burtenshaw committed:
feat: implement ingestion mapping as class

burtenshaw committed:
feat: use ingestion mapping class in dataset records not dataset records

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
chore: tidy imports

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' of https://github.com/argilla-io/argilla into spike/mapping-to-tuple

burtenshaw committed:
docs: update mapping parameters in how to guides

burtenshaw committed:
test: broaden suggestion mapping in test

burtenshaw committed:
feat: extract dot notation with regex not string splitting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon commented on argilla/docs/how_to_guides/record.md:
````diff
@@ -318,24 +318,31 @@ Suggestions refer to suggested responses (e.g. model predictions) that you can a
 
     ```python
-    # Add records to the dataset with the label 'my_label'
+        # Add records to the dataset with the label 'my_label'
````
```suggestion
        # Add records to the dataset with the label question 'my_label'
```

frascuchon commented on argilla/src/argilla/records/_dataset_records.py:
```diff
@@ -357,26 +357,36 @@ def to_datasets(self) -> HFDataset:
+        record_mapper = IngestedRecordMapper(mapping=mapping, dataset=self.__dataset, user_id=user_id)
+        for record in records:
+            try:
+                if not isinstance(record, Record):
```
we can avoid unexpected errors trying to map some weird content.
```suggestion
                if isinstance(record, dict):
```

frascuchon commented on argilla/src/argilla/records/_mapping.py:
```diff
@@ -0,0 +1,350 @@
+        # Get the available attributes, types, and parameters from the schema.
+        available_attributes = list(self._schema.keys()) + ["id"]
+        available_types = [type_.value for type_ in AttributeType]
+        available_parameters = [param.value for param in ParameterType]
```
Maybe we can move this list comprehension to the corresponding classes and define them as properties:
```python
class AtributeType(..):
   ...
   @property
   def values(self):
         return [t.value for t in self]

burtenshaw replied:
I can't implement them as properties without needing to init the class to access, so I've implemented them as classmethods instead.

Here's the attempt with properties:

```python
from enum import Enum
from typing import List

class ParameterType(str, Enum):
    """Parameter types are the different 'sub' values of a records attribute.
    For example, the value, score, or agent of a suggestion."""

    VALUE = "value"
    SCORE = "score"
    AGENT = "agent"
    
    @property
    def values(self) -> List[str]:
        return [param.value for param in ParameterType]

class AttributeType(str, Enum):
    """Attribute types are the different types of attributes a record can have."""

    FIELD = "field"
    SUGGESTION = "suggestion"
    RESPONSE = "response"
    METADATA = "metadata"
    VECTOR = "vector"
    ID = "id"
    
    @property
    def values(self) -> List[str]:
        return [attr.value for attr in AttributeType]

# Example usage:
print(ParameterType.VALUE.values)
print(AttributeType.FIELD.values)
```

frascuchon commented on argilla/src/argilla/records/_mapping.py:
```diff
@@ -0,0 +1,350 @@
+    field: Dict[str, AttributeRoute]
+    metadata: Dict[str, AttributeRoute]
+    vector: Dict[str, AttributeRoute]
+    id: Dict[str, AttributeRoute]
```
Do we support multiple ID definitions in mapping? I would expect here a single `AttributeRoute` for id.
```python
   id: AttributeRoute
```

frascuchon commented on argilla/src/argilla/records/_mapping.py:
```diff
@@ -0,0 +1,350 @@
+
+        """
+
+        record_id = data.get(self.mapping.id["id"].source)
```
Could we do `self.mapping.id.source` instead?

frascuchon commented on argilla/src/argilla/records/_mapping.py:
```diff
@@ -0,0 +1,350 @@
+        elif attribute.name == "id":
+            attribute.type = AttributeType.ID
+        else:
+            warnings.warn(message=f"Record attribute {attribute.name} is not in the schema or mapping so skipping.")
```
In which scenario this will happen? Would be safe to return the attribute for those cases?

burtenshaw replied:
This if statement is true if the use has mapped to a field name and that field name appears in the schema.

frascuchon replied:
In that case, the warning message is confusing:
```python
            warnings.warn(message=f"Record attribute {attribute.name} is not in the schema or mapping so skipping.")
```

burtenshaw replied:
I've replaced the warning with an error because in fact this should only happen with the user does something in valid. I've also changed the message.

frascuchon commented on argilla/src/argilla/records/_mapping.py:
```diff
@@ -0,0 +1,350 @@
+                )
+
+            elif not mapping.suggestion[question.name].has_value:
+                mapping.suggestion[question.name].parameters.append(AttributeParameter(source=question.name))
```
which mapping will result in this? Should we avoid mappings without values instead?

burtenshaw replied:
This is used when an attributes are mapped accross multiple columns. i.e. a suggestion's score, agent, and value:

```python
mock_mapping = {
        "my_prompt": "prompt",
        "my_label": "label.suggestion.value",
        "score": "label.suggestion.score",
        "model": "label.suggestion.agent",
    }
```

frascuchon replied:
But for that case, the mapping.suggestion[question.name].has_value will return True, since the "default" mapping is computed after the first mapping normalization, right?

burtenshaw replied:
I don't think this is relevant with the latest changes? to merge your PR.

frascuchon commented on argilla/src/argilla/records/_mapping.py:
```diff
@@ -0,0 +1,350 @@
+            warnings.warn(message=f"Record attribute {attribute.name} is not in the schema or mapping so skipping.")
+        return attribute
+
+    def _schematize_default_attributes(self, mapping: RecordAttributesMap) -> RecordAttributesMap:
```
Not sure because I'm missing some scenarios, but maybe  would be easier if we first build the "default" RecordAttributesMap from the schema and then update/override with content provided in `mapping`?
```python
def __init__(...):
        default_mapping = self.._schematize_default_attributes(schema=dataset.schema)
        self.mapping = self._schematize_default_attributes(default=default_mapping, mapping or {})
```

burtenshaw replied:
I can see why that make sense, and the default method is easier to implement, but I think this is easier to follow.

I found that the `_schematize_mapped_attributes` became complex and needed to pick apart the `AttributeRoute` object to edit and add mapped attributes and sub attributes to the the routes.

burtenshaw committed:
docs: typo in docs

burtenshaw committed:
feat: improve record switch in ingest method

burtenshaw committed:
feat: refactor id mapping away from dict to type

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
feat: use class methods for type and parameter values

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' of https://github.com/argilla-io/argilla into spike/mapping-to-tuple

frascuchon committed:
[REFACTOR] generate default mapping and extends it with custom mapping dict (#5151)

<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

Closes #<issue_number>

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Refactor (change restructuring the codebase without changing
functionality)

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- follows the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

burtenshaw committed:
refactor: migrate mapper into module from file

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
fix: raise error not warn when mapped attribute is unknown

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' of https://github.com/argilla-io/argilla into spike/mapping-to-tuple

frascuchon committed:
Merge branch 'develop' into spike/mapping-to-tuple
